readthedocs/vcs_support/backends/git.py

```diff
@@ -523,8 +537,84 @@ def commit(self):
+        we parse the .gitmodules file. For this we make use of the
+        ``git config --get-regexp`` command.
+
+        Keys and values from the config can contain spaces,
+        in order to parse the output unambiguously, we use the
+        ``--null`` option to separate each result with a null character,
+        and each key and value with a newline character.
+
+        The command will produce an output like this:
+
+        .. code-block:: text
+
+           submodule.submodule-1.url\nhttps://github.com/\0
+           submodule.submodule-1.path\nsubmodule-1\0
+           submodule.submodule-2.path\nsubmodule-2\0
+           submodule.submodule-2.url\nhttps://github.com/\0
+           submodule.submodule-3.url\nhttps://github.com/\0
+           submodule.submodule-4.path\n\0
+
+        .. note::
+
+           - In the example each result is put in a new line for readability.
+           - Isn't guaranteed that the url and path keys will appear next to each other.
+           - Isn't guaranteed that that all submodules will have a url and path.
+
+        """
+        exit_code, stdout, _ = self.run(
+            "git",
+            "config",
+            "--null",
+            "--file",
+            ".gitmodules",
+            "--get-regexp",
+            # Get only the URL and path keys of each submodule.
+            r"^submodule\..*\.(url|path)$",
+            record=False,
+        )
+        if exit_code != 0:
+            return []
+
+        # Group the URLs and paths by submodule name/key.
+        submodules = {}
+        keys_and_values = stdout.split("\0")
+        for key_and_value in keys_and_values:
+            try:
+                key, value = key_and_value.split("\n", maxsplit=1)
+            except ValueError:
+                # This should never happen, but we log a warning just in case
+                # Git doesn't return the expected format.
+                log.warning("Wrong key and value format.", key_and_value=key_and_value)
+                continue
+
+            if key.endswith(".url"):
+                key = key.removesuffix(".url")
+                submodules.setdefault(key, {})["url"] = value
+            elif key.endswith(".path"):
+                key = key.removesuffix(".path")
+                submodules.setdefault(key, {})["path"] = value
+            else:
+                # This should never happen, but we log a warning just in case the regex is wrong.
+                log.warning("Unexpected key extracted fom .gitmodules.", key=key)
+
+        for submodule in submodules.values():
+            # If the submodule doesn't have a URL or path, we skip it,
+            # since it's not a valid submodule.
+            url = submodule.get("url")
+            path = submodule.get("path")
+            if not url or not path:
+                continue
+            yield GitSubmodule(
+                url=url,
+                path=path,
+            )
```

Comment:
Why are you returning an generator instead of just a regular list? The logic is pretty simple to be calculated upfront.

Reply:
So we don't return the whole list when checking if the project has at least one submodule

https://github.com/readthedocs/readthedocs.org/blob/d8a36079389db5c6fc36cd428c957b9e1ecca3f4/readthedocs/vcs_support/backends/git.py#L296-L296